src/iotjs_module_httpparser.cpp

```diff
@@ -0,0 +1,483 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include "iotjs_def.h"
+#include "iotjs_module_httpparser.h"
+#include "http_parser.h"
+#include "iotjs_objectwrap.h"
+#include "iotjs_module_buffer.h"
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+
+namespace iotjs {
+
+
+#define JSETPROPERTY( container_, fname_, vname_ )                      \
+  do {                                                                  \
+    JObject jobj(vname_ );                                              \
+    container_.SetProperty(fname_, jobj);                               \
+  } while(0)
+
+
+// max header size. If header field > HEADER_MAX, flush header to JS side.
+// increse this to minimize inter JS-C call
+#define HEADER_MAX 10
+
+
+
+class HTTPParserWrap : public JObjectWrap {
+public:
+  explicit HTTPParserWrap(JObject& parser_, http_parser_type type)
+    : JObjectWrap(parser_) {
+    Initialize(type);
+    parser.data = this;
+  }
+
+  ~HTTPParserWrap() {
+  }
+
+  void Initialize(http_parser_type type);
+
+  // http-parser callbacks
+  int OnMessageBegin() {
+    url.MakeEmpty();
+    status_msg.MakeEmpty();
+    return 0;
+  }
+
+  int OnUrl(const char* at, size_t length) {
+    url.Append(at, length);
+    return 0;
+  }
+
+  int OnStatus(const char* at, size_t length) {
+    status_msg.Append(at, length);
+    return 0;
+  }
+
+  int OnHeaderField(const char* at, size_t length) {
+
+    if (n_fields == n_values) {
+      n_fields++;
+      if (n_fields == HEADER_MAX) {
+        Flush(); // to JS world
+        n_fields = 1;
+        n_values = 0;
+      }
+      fields[n_fields-1].MakeEmpty();
+    }
+    IOTJS_ASSERT(n_fields == n_values + 1);
+    fields[n_fields-1].Append(at, length);
+
+    return 0;
+  }
+
+  int OnHeaderValue(const char* at, size_t length) {
+    if (n_fields != n_values) {
+      n_values++;
+      values[n_values-1].MakeEmpty();
+    }
+
+    IOTJS_ASSERT(n_fields == n_values);
+
+    values[n_values-1].Append(at, length);
+
+    return 0;
+  }
+
+  JObject makeHeader(){
+    // FIXME: This must be impl. as JS Array
+    JObject jheader;
+    JSETPROPERTY(jheader, "length", (n_fields)*2);
+    for (int i=0;i<n_fields;i++) {
+      char index_string0 [5];
+      char index_string1 [5];
+      sprintf(index_string0,"%d",i*2);
+      sprintf(index_string1,"%d",i*2+1);
+      JObject v(values[i].data());
+      JObject f(fields[i].data());
+      jheader.SetProperty(index_string0, f);
+      jheader.SetProperty(index_string1, v);
+
+    }
+    return jheader;
+  }
+
+
+  int OnHeadersComplete() {
+
+    JObject jobj = jobject();
+    JObject func = jobj.GetProperty("OnHeadersComplete");
+    IOTJS_ASSERT(func.IsFunction());
+
+    // URL
+    JArgList argv(1);
+    JObject info;
+
+    // HEADERS
+    if (flushed) {
+      Flush();
```

Comment:
In previous ```Flush()``` call, ```url``` was already flushed to JS side. However, if not ```flushed```, all headers and ```url``` start to flush here.